docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,473 @@
+  sync manager knows via some magic how to obtain these handles.
+
+* Some apps will use a combination of Rust components and "legacy"
+  engines. For example, iOS is moving some of the engines to using Rust
+  components, while other engines will be ported after delivery of the
+  sync manager, if they are ported at all. We also plan
+  to introduce some rust engines into desktop without integrating the
+  "sync manager"
+
+* The rust components themselves are designed to be consumed as individual
+  components - the "logins" component doesn't know anything about the
+  "bookmarks" component.
+
+There are a couple of gotchyas in the current implementations too - there's an
+issue when certain engines don't yet appear in meta/global - see bug 1479929
+for all the details.
+
+The tl;dr of the above is that each rust component should be capable of
+working with different sync managers. That said though, let's not over-engineer
+this and pretend we can design a single, canonical thing that will not need
+changing as we consider desktop and iOS.
+
+## State, state and more state. And then some state.
+
+There's loads of state here. The app itself has some state. The high-level
+Sync Manager component will have state, the low-level component will have state,
+and each engine has state. Some of this state will need to be persisted (either
+on the device or on the storage servers) and some of this state can be considered
+ephemeral and lives only as long as the app.
+
+A key challenge will be defining this state in a coherent way with clear
+boundaries between them, in an effort to allow decoupling of the various bits
+so Desktop and iOS can fit into this world.
+
+This state management should also provide the correct degree of isolation for
+the various components. For example, each engine should only know about state
+which directly impacts how it works. For example, the keys used to encrypt
+a collection should only be exposed to that specific engine, and there's no
+need for one engine to know what info/collections returns for other engines,
+nor whether the device is currently connected to WiFi.
+
+A thorn here is for persisted state - it would be ideal if the low-level
+component could avoid needing to persist any state, so it can avoid any
+kind of storage abstraction. A solution here might be that the low-level
+component can delegate state storage to the high-level component in an
+opaque way.
+
+# Implementation plan for the low-level component.
+
+Let's try and move into actionable decisions for the implementation. We expect
+the implementation of the low-level component to happen first, followed very
+closely by the implementation of the high-level component for Android. So we
+focus first on these.
+
+## Clients Engine
+
+The clients engine includes some meta-data about each client. We've decided
+we can't replace the clients engine with the FxA device record and we can't
```

Comment:
"Because the ID to too ephemeral"? :) As above, I'm rethinking that - it's a real PITA for client code to deal with them being different, so we should suck it up and see what breaks :)